client/executor/src/async_externalities.rs

```diff
@@ -0,0 +1,214 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Async externalities.
+
+use std::{
+	any::{TypeId, Any},
+	sync::Arc,
+};
+use sp_core::{
+	storage::{ChildInfo, TrackedStorageKey},
+	traits::{Externalities, SpawnNamed, TaskExecutorExt},
+};
+use sp_io::RuntimeSpawnExt;
+use sp_externalities::{Extensions, ExternalitiesExt as _};
+use crate::native_executor::RuntimeInstanceSpawn;
+
+/// Simple state-less externalities for use in async context.
+///
+/// Will panic if anything is accessing the storage.
+#[derive(Debug)]
+pub struct AsyncExternalities {
+    extensions: Extensions,
+}
+
+pub fn new_async_externalities(
+	scheduler: Box<dyn SpawnNamed>,
+	module: Arc<dyn sc_executor_common::wasm_runtime::WasmModule>,
+) -> Result<AsyncExternalities, &'static str> {
+	let extensions = Extensions::default();
+	let mut res = AsyncExternalities { extensions };
+	let mut ext = &mut res as &mut dyn Externalities;
+	ext.register_extension::<TaskExecutorExt>(TaskExecutorExt(scheduler.clone()))
+		.map_err(|_| "Failed to register task executor extension.")?;
+	ext.register_extension::<RuntimeSpawnExt>(
+		RuntimeSpawnExt(Box::new(
+			RuntimeInstanceSpawn::new(module, scheduler)
+		))
+	)
+	.map_err(|_| "Failed to register task executor extension.")?;
+
+	Ok(res)
+}
+
+type StorageKey = Vec<u8>;
+
+type StorageValue = Vec<u8>;
+
+impl Externalities for AsyncExternalities {
```

Comment:
This part could be much simpler if we refactored storage access and stuff that is now in extensions into capability-based externalities.